sdk/keyvault/azure-keyvault-keys/src/main/java/com/azure/security/keyvault/keys/cryptography/CryptographyAsyncClient.java

```diff
@@ -0,0 +1,514 @@
+     * operation is supported for both symmetric keys and asymmetric keys. In case of asymmetric keys public portion of the key is used
+     * for encryption. This operation requires the keys/encrypt permission.
+     *
+     * <p>The {@link EncryptionAlgorithm encryption algorithm} indicates the type of algorithm to use for decrypting the specified encrypted content. Possible values
+     * for assymetric keys include: {@link EncryptionAlgorithm#RSA1_5 RSA1_5}, {@link EncryptionAlgorithm#RSA_OAEP RSA_OAEP} and {@link EncryptionAlgorithm#RSA_OAEP_256 RSA_OAEP_256}.
+     * Possible values for symmetric keys include: {@link EncryptionAlgorithm#A128CBC A128CBC}, {@link EncryptionAlgorithm#A128CBC_HS256 A128CBC-HS256},
+     * {@link EncryptionAlgorithm#A192CBC A192CBC}, {@link EncryptionAlgorithm#A192CBC_HS384 A192CBC-HS384}, {@link EncryptionAlgorithm#A256CBC A256CBC} and {@link EncryptionAlgorithm#A256CBC_HS512 A256CBC-HS512} </p>
+     *
+     * @param algorithm The algorithm to be used for encryption.
+     * @param plaintext The content to be encrypted.
+     * @param iv The initialization vector
+     * @param authenticationData The authentication data
+     * @throws ResourceNotFoundException if the key cannot be found for encryption.
+     * @return A {@link Mono} containing a {@link EncryptResult} whose {@link EncryptResult#cipherText() cipher text} contains the encrypted content.
+     */
+    public Mono<EncryptResult> encrypt(EncryptionAlgorithm algorithm, byte[] plaintext, byte[] iv, byte[] authenticationData) {
+        return withContext(context -> encrypt(algorithm, plaintext, context, iv, authenticationData));
+    }
+
+    Mono<EncryptResult> encrypt(EncryptionAlgorithm algorithm, byte[] plaintext, Context context, byte[] iv, byte[] authenticationData) {
+        Objects.requireNonNull(algorithm);
+
+        if(key == null){
+            this.key = getKey().block().value().keyMaterial();
+        }
+
+        if (!checkKeyPermissions(this.key.keyOps(), KeyOperation.ENCRYPT)){
+            return Mono.error(new UnsupportedOperationException(String.format("Encrypt Operation is not supported for key with id %s", key.kid())));
+        }
+
+            switch(key.kty()){
+            case RSA:
+            case RSA_HSM:
+                return rsaKeyCryptographyClient.encryptAsync(algorithm, plaintext, iv, authenticationData, context, key);
+            case EC:
+            case EC_HSM:
+                return ecKeyCryptographyClient.encryptAsync(algorithm, plaintext, iv, authenticationData, context, key);
+            case OCT:
+                return symmetricKeyCryptographyClient.encryptAsync(algorithm, plaintext, iv, authenticationData, context, key);
+            default:
+                throw new UnsupportedOperationException(String.format("Encrypt Async is not allowed for Key Type: %s", key.kty().toString()));
+        }
```

Comment:
You might consider creating a common abstract base class for all the cryptography client implementations. Then you could just have field, something like implCryptoClient, and avoid having this switch statement in every method.

Reply:
Yeah, I started doing this refactor in the morning. will push soon.

Reply:
This refactor is complete